control-plane/api-gateway/gatekeeper/dataplane.go

```diff
@@ -0,0 +1,172 @@
+)
+
+const (
+	consulDataplaneDNSBindHost   = "127.0.0.1"
+	consulDataplaneDNSBindPort   = 8600
+	sidecarUserAndGroupID        = 5995
+	defaultPrometheusScrapePath  = "/metrics"
+	defaultEnvoyProxyConcurrency = 1
+	volumeName                   = "consul-connect-inject-data"
+)
+
+func consulDataplaneContainer(config apigateway.HelmConfig, name, namespace string) (corev1.Container, error) {
+	// Extract the service account token's volume mount.
+	var (
+		err             error
+		bearerTokenFile string
+	)
+
+	if config.AuthMethod != "" {
+		bearerTokenFile = "/var/run/secrets/kubernetes.io/serviceaccount/token"
+	}
+
+	args, err := getDataplaneArgs(namespace, config, bearerTokenFile, name)
+	if err != nil {
+		return corev1.Container{}, err
+	}
+
+	probe := &corev1.Probe{
+		ProbeHandler: corev1.ProbeHandler{
+			HTTPGet: &corev1.HTTPGetAction{
+				Port: intstr.FromInt(constants.ProxyDefaultHealthPort),
+				Path: "/ready",
+			},
+		},
+		InitialDelaySeconds: 1,
+	}
+
+	container := corev1.Container{
+		Name:  name,
+		Image: config.ImageDataplane,
+
+		// We need to set tmp dir to an ephemeral volume that we're mounting so that
+		// consul-dataplane can write files to it. Otherwise, it wouldn't be able to
+		// because we set file system to be read-only.
+		Env: []corev1.EnvVar{
+			{
+				Name:  "TMPDIR",
+				Value: "/consul/connect-inject",
+			},
+			{
+				Name: "NODE_NAME",
+				ValueFrom: &corev1.EnvVarSource{
+					FieldRef: &corev1.ObjectFieldSelector{
+						FieldPath: "spec.nodeName",
+					},
+				},
+			},
+			{
+				Name:  "DP_SERVICE_NODE_NAME",
+				Value: "$(NODE_NAME)-virtual",
+			},
+		},
+		VolumeMounts: []corev1.VolumeMount{
+			{
+				Name:      volumeName,
+				MountPath: "/consul/connect-inject",
+			},
+		},
+		Args:           args,
+		ReadinessProbe: probe,
+	}
+
+	// Configure the Readiness Address for the proxy's health check to be the Pod IP.
+	container.Env = append(container.Env, corev1.EnvVar{
+		Name: "DP_ENVOY_READY_BIND_ADDRESS",
+		ValueFrom: &corev1.EnvVarSource{
+			FieldRef: &corev1.ObjectFieldSelector{FieldPath: "status.podIP"},
+		},
+	})
+	// Configure the port on which the readiness probe will query the proxy for its health.
+	container.Ports = append(container.Ports, corev1.ContainerPort{
+		Name:          "proxy-health",
+		ContainerPort: int32(constants.ProxyDefaultHealthPort),
+	})
+
+	// If not running in an OpenShift environment,
+	// skip setting the security context and let OpenShift set it for us.
+	if !config.EnableOpenShift {
+		container.SecurityContext = &corev1.SecurityContext{
+			RunAsUser:              pointer.Int64(sidecarUserAndGroupID),
+			RunAsGroup:             pointer.Int64(sidecarUserAndGroupID),
+			RunAsNonRoot:           pointer.Bool(true),
+			ReadOnlyRootFilesystem: pointer.Bool(true),
+		}
+	}
+
+	return container, nil
+}
+
+func getDataplaneArgs(namespace string, config apigateway.HelmConfig, bearerTokenFile string, name string) ([]string, error) {
```

Comment:
This is definitely a fine approach. However, we can also set these values as environment variables in the container and that may lead to a more maintainable method of setting these. 

That can be a post-feature-freeze work item if we want to make the change.

Reply:
👍 I'm fine with either the args or env variable approach, just did this since I could base it off existing code and didn't have to put as much thought behind how to implement it at first 😅